### DISCUSSION TITLE
symbol's function definition is void use-package.

### DISCUSSION THREAD

**Discussion by yigitako** [minad/corfu#412] [Q&A] [answered]
Greetings,
I have successfully installed the **Corfu package** with **package-install**. Then, I opened  a new buffer with a Python script. Afterwards, I executed the command M-x global-corfu-mode RET, and it returned the confirmation that the global-corfu-mode is now enabled, but the autocompletion feature didn't worked.

So, Following this, I added the example configuration to my initialization file and proceeded to restart Emacs. However, upon restarting, Emacs generated an error stating, "symbol's function definition is void use-package."

To address this issue, I have taken the following steps:
-- I have added (require 'corfu) at the beginning of the config. Didn't work and throw the same error
-- I have verified that the corfu is installed correctly by checking the package-list-pacakges.
**corfu              1.1           installed             COmpletion in Region FUnction**
-- Restarted emascs several times.

***-- TECHNICAL INFO***
**EMACS VERSION** :: GNU Emacs 27.1 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.33, cairo version 1.16.0)
 of 2023-08-24, modified by Debian

**OS** ::  Trisquel GNU/Linux Aramo (11.0) x86_64](url)

**Kernel** :: 5.15.0-91-generic 
--  EMACS init.el -- 
```
(use-package corfu
  ;; Optional customizations
  ;; :custom
  ;; (corfu-cycle t)                ;; Enable cycling for `corfu-next/previous'
  (corfu-auto t)                 ;; Enable auto completion
  ;; (corfu-separator ?\s)          ;; Orderless field separator
  ;; (corfu-quit-at-boundary nil)   ;; Never quit at completion boundary
  ;; (corfu-quit-no-match nil)      ;; Never quit, even if there is no match
  ;; (corfu-preview-current nil)    ;; Disable current candidate preview
  ;; (corfu-preselect 'prompt)      ;; Preselect the prompt
  ;; (corfu-on-exact-match nil)     ;; Configure handling of exact matches
  ;; (corfu-scroll-margin 5)        ;; Use scroll margin

  ;; Enable Corfu only for certain modes.
  ;; :hook ((prog-mode . corfu-mode)
  ;;        (shell-mode . corfu-mode)
  ;;        (eshell-mode . corfu-mode))

  ;; Recommended: Enable Corfu globally.  This is recommended since Dabbrev can
  ;; be used globally (M-/).  See also the customization variable
  ;; `global-corfu-modes' to exclude certain modes.
  :init
  (global-corfu-mode))

;; A few more useful configurations...
(use-package emacs
  :init
  ;; TAB cycle if there are only few candidates
  (setq completion-cycle-threshold 3)

  ;; Emacs 28: Hide commands in M-x which do not apply to the current mode.
  ;; Corfu commands are hidden, since they are not supposed to be used via M-x.
  ;; (setq read-extended-command-predicate
  ;;       #'command-completion-default-include-p)

  ;; Enable indentation+completion using the TAB key.
  ;; `completion-at-point' is often bound to M-TAB.
  (setq tab-always-indent 'complete))
```

**Comment by minad** [comment, marked as answer]
You should first install `use-package` from [GNU ELPA](https://elpa.gnu.org/packages/use-package.html).

**Reply by yigitako** [reply]
Thank you for your assistance and time, CORFU works without any problem.Also,  I'd like to emphasize that when a package, such as 'use-package,' is mentioned, not providing a information may lead to the assumption that it is a built-in tool. To avoid confusion,  you can add "first install use-package from GNU ELPA." to documentation.

**Reply by minad** [reply]
`use-package` is a builtin, but only since Emacs 29. However it is a *very common* package, so it is reasonable to expect familiarity with it and to not give many details about it. Unfortunately there is a learning curve and barrier for new users here, which is hard to avoid, if we don't want to expand the documentation infinitely. As a new user one either starts small with bare Emacs, learning everything step by step, including `use-package`, or one starts with a starter kit which already includes `use-package` and maybe Vertico or Corfu.
